site/docs/expressions/scalar_functions.md

```diff
@@ -1,52 +1,118 @@
 # Scalar Functions
 
-Scalar functions are a function that takes in values from a single record and produces an output value. To clearly specify the definition of functions, Substrait declares a extensible specification plus binding approach to function resolution. This ensures a clear definition of function behavior separate from a particular use of each function while avoiding operations 
+Scalar functions are a function that takes in values from a single record and produces an output value. To clearly specify the definition of functions, Substrait declares a extensible specification plus binding approach to function resolution. A scalar function signature includes the following properties:
 
-Substrait supports a number of functions. A scalar function signature includes the following properties:
+| Property               | Description                                                  | Required                            |
+| ---------------------- | ------------------------------------------------------------ | ----------------------------------- |
+| Name                   | One or more user friendly utf8 strings that are used to reference this function in languages | At least one value is required.     |
+| List of arguments      | Argument properties are defined below. Arguments can be fully defined, parameterized or wildcarded. See further details below. | Optional, defaults to niladic.      |
+| Deterministic          | Whether this function is expected to reproduce the same output when it is invoked multiple times with the same input. This informs a plan consumer on whether it can constant reduce the defined function. An example would be a random() function, which is typically expected to be evaluated repeatedly despite having the same set of inputs. | Optional, defaults to true.         |
+| Session Dependent      | Whether this function is influenced by the session context it is invoked within. For example, a function may be influenced by a user who is invoking the function, the time zone of a session, or some other non-obvious parameter. This can inform caching systems on whether a particular function is cacheable. | Optional, defaults to false.        |
```

Comment:
Naive question.  Do we really need both session dependent and deterministic or can session dependent just be one particular case of non-deterministic?  Will a system be able to optimize more efficiently knowing a function is deterministic but not session dependent?

Reply:
They play in different spaces/influence different optimization opportunities. 

Deterministic describes whether or not a value could be reduced to a constant in some expressions. If it is non-deterministic, you must execute for each record. Select rand(), a,b,c from foo is a good example. You wouldn't expect rand() to be the same value on any record. If you did the same thing as add(1,1), you could replace with the literal 2.

Session dependent describes whether or not a plan's interpretation is impacted by session properties and thus means you can't share something like a plan or data cache between different sessions (without also keying session properties). A function that returns username is a prime example.